ckanext/security/fanstatic/login_ajax.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+(function ($) {
+  var getLoginForm = function () {
+    return $('#mfa-login-form')
+  }
+
+  var showQRCode = function (loginState) {
+    $('#totp-secret').text(loginState.totpSecret)
+    $('#mfa-setup').show()
+
+    new QRious({
+      size: 250,
+      element: $('#qr-code-container')[0],
+      value: loginState.totpChallengerURI
+    })
+  }
+
+  var handleMfaValidation = function (validCode) {
+    if (validCode) {
+      getLoginForm().off('submit', validateLoginCredentials)
+      getLoginForm().submit()
+    } else {
+      showError('mfa')
+    }
+  }
+
+  var validateLoginCredentials = function (e) {
+    e.preventDefault()
+
+    var form = $(this)
+    var buttons = $('[type=submit]')
+    buttons.prop('disabled', true)
+
+    $.ajax({
+      url: '/mfa_login',
+      method: 'POST',
+      data: form.serialize(),
+      success: function (loginState) {
+        hideError()
+
+        if (loginState.hasOwnProperty('mfaCodeValid')) {
+          handleMfaValidation(loginState.mfaCodeValid)
+          return
+        }
```

Comment:
First form submit will validate the username and password only, so we show the 2fa form if it succeeds. The next submit has the username, password and 2fa code, and if that all succeeds we detach the form submit handler and do a standard form post, completing the login.

Reply:
Yes, that was the main reason... if they submit the first username/password without the 2fa code, the authentication mechanism will fail as it requires all 3 to succeed. The two step login was preferred rather than all in the same form, this seemed like a relatively simple way to provide error/validation feedback for username and password, and then the 2fa code afterwards.